Initial multi-drive command

yyoon opened the pull request.

kaiyuanw committed:
Enable develop mode

kaiyuanw committed:
Add multi_drive.dart to support multi-device application driver

kaiyuanw committed:
Add default debug-ports for debugging

kaiyuanw committed:
Modified executable.dart to make command line tool support 'flutter multi-drive ...'

kaiyuanw committed:
Delete default debug-ports since that will not be used.  Everything should be specified in the specs file

kaiyuanw committed:
Override toString() in AndroidDevice for debugging

kaiyuanw committed:
Modified MultiDriveCommand to make it support loading specs files instead of reading command line options

kaiyuanw committed:
Build initial command line tool that support loading specs file and run applications on multiple devices

kaiyuanw committed:
Modified MultiDriveCommand class and make it extends FlutterCommand so the target flag is not required anymore

kaiyuanw committed:
Remove unnecessary code and add some documentations

kaiyuanw committed:
Revert AndroidDevice toString() method and add null checking for the specs argument in MultiDriveCommand

yyoon commented on bin/flutter:
```diff
@@ -41,6 +41,8 @@ if [ ! -f "$SNAPSHOT_PATH" ] || [ ! -f "$STAMP_PATH" ] || [ `cat "$STAMP_PATH"`
   echo $REVISION > "$STAMP_PATH"
 fi
 
+FLUTTER_DEV=1
```
Please remove.
(Even if you want to use this locally, you shouldn't be committing this.)

kaiyuanw replied:
Solved.

yyoon commented on packages/flutter_tools/lib/src/android/android_device.dart:
```diff
@@ -52,6 +52,17 @@ class AndroidDevice extends Device {
+    "modelID: $modelID,\n"
+    "deviceCodeName: $deviceCodeName\n";
+  */
+
```
Please remove this, if this is something you added for debugging purposes. Let's try our best to not modify the existing flutter code.

Also, in general, it is not a good practice to check in commented out code, unless you have a strong reason for doing so.

kaiyuanw replied:
Solved.  Thanks for pointing out.

yyoon commented on packages/flutter_tools/lib/src/base/common.dart:
```diff
@@ -5,3 +5,4 @@
 const int kDefaultObservatoryPort = 8100;
 const int kDefaultDiagnosticPort  = 8101;
 const int kDefaultDrivePort       = 8183;
+// const String kDefaultMultiDrivePort = '8183,8184';
```
Please remove.

kaiyuanw replied:
Removed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+    /*
+    argParser.addOption('route',
+        help: 'Which route to load when running the app.');
+    */
```
Please remove the commented out block.
We might want to add the 'route' option in the spec file later.

kaiyuanw replied:
'route' option removed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+      defaultsTo: null,
+      allowMultiple: false,
+      help:
+        'Config file that specifies the devices, apps and debug-ports for testing.'
```
'Config file...' --> 'Path to the config file'

Also, wrap the line at 80 characters, as done in the other flags above.
(this applies to the entire file)

kaiyuanw replied:
Thanks.  Rephrased.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+    //   allowMultiple: true,
+    //   splitCommas: true,
+    //   help: 'Listen to a list of ports for a debug connection.'
+    // );
```
Please remove.

kaiyuanw replied:
Comments removed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+
+  bool get traceStartup => argResults['trace-startup'];
+
+  String get route => null;//argResults['route'];
```
Let's add a TODO comment and make it `'/'` instead of null. So the resulting code should look like:

``` dart
// TODO(kaiyuanw): read the route for each device from the specs file.
String get route => '/';
```

kaiyuanw replied:
The default value for 'trace-startup' in the flutter source code is null, so it's better to conform with the flutter team if we use the route argument.  We may change this later if it turns out that we need to specify this argument to non-null value.  Thanks for pointing this out.

yyoon replied:
Keeping the value as `null` is fine, but you should still add a `TODO` here.

kaiyuanw replied:
Done.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+  final String name = 'multi-drive';
+
+  @override
+  final String description = 'Runs Flutter Multi-Device Driver tests for the current project.';
```
'Runs' -> 'Run'

kaiyuanw replied:
Fixed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+
+  Future<dynamic> _loadSpecs(String specsPath) async {
+    // Read specs file into json format
+    dynamic spec = JSON.decode(await new io.File(specsPath).readAsString());
```
1. `spec` -> `specs`, just to be consistent.
2. Some error handling should be done here. What if the file doesn't exist, for example?

kaiyuanw replied:
The spec is renamed to newSpecs to solve the confusion.  Also, exception/error handling is added.  You might want to review the exception/error handling part.  Thank you.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+    this.specs = await _loadSpecs(specsPath);
+    print(specs);
+
+    this._devices = await targetDevicesFinder();
```
Why not just call `findTargetDevices()`?

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+TargetDevicesFinder targetDevicesFinder = findTargetDevices;
+void restoreTargetDevicesFinder() {
+  targetDevicesFinder = findTargetDevices;
+}
```
I see. For now, having the `TargetDevicesFinder` and `restoreTargetDevicesFinder()` seems unnecessary, but these would be useful if you want to write some tests for this multi-drive command.

You should write some tests, similar to `flutter_tools/test/drive_test.dart`.

kaiyuanw replied:
Thanks for pointing this out.  I'll write test once I think all function signatures are stable.  That will happen soon.  For now, I'll leave a todo list in the comments.  Will solve this problem soon.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+  if (deviceManager.hasSpecifiedDeviceId) {
+    return deviceManager.getDeviceById(deviceManager.specifiedDeviceId);
+  }
+  */
```
Please remove.

kaiyuanw replied:
Removed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+
+Future<int> startMultiDeviceApps(MultiDriveCommand command) async {
+  // Load all device ids
+  Map<String, dynamic> devices = command.specs['devices'];
```
The variable name `devices` seems overloaded. Here, it means the _devices in the specs_, but in other places, `devices` just means all the connected/available devices. You should be distinguishing those.

I suggest calling this variable `specifiedDevices` instead.

kaiyuanw replied:
Variable renamed to `specifiedDevices`

yyoon replied:
I don't see this change. Maybe you forgot to commit?

kaiyuanw replied:
I was confused and changed the `devices` variable in the `findDevice` function.  Now fixed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+    }
+
+    printTrace('Stopping previously running application, if any.');
+    await appsStopper(command.applicationPackages, device);
```
This line seems weird to me. How does it know the right package name to use for this device?
`command.applicationPackages` seems to be inferring the package name from the current directory, which is probably not what you want.

kaiyuanw replied:
This is a limitation of the current implementation.  Currently, `flutter multi-drive` command only support running multiple main() functions in a single application root folder.  I'll revisit this issue and solve it later.  Right now I'll leave some comments in the code.  Good catch!!!

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+    printTrace('Stopping previously running application, if any.');
+    await appsStopper(command.applicationPackages, device);
+
+    installAndStartAppFunctions.add(makeInstallAndStartApp(command, device, mainPath, config['debug-port']));
```
Please wrap this line.

kaiyuanw replied:
Done.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+  // });
+  // Install and start apps in parallel
+  // TODO: modify internal synchronous calls to asynchronous calls
+  Future.wait(installAndStartAppFunctions)
```
Please remove the irrelevant comments.
If you want to keep the TODO comment (I think you should), the correct format should be:
`// TODO(<your ID>): <some stuffs to do later>`

kaiyuanw replied:
Irrelevant comments deleted.  Comments reformatted.  Thanks.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+  String debugPort) async {
+    printTrace('Installing application package.');
+    ApplicationPackage package = command.applicationPackages
+        .getPackageForPlatform(device.platform);
```
Again, please make sure that this package name is correctly obtained.

kaiyuanw replied:
Same problem.  Will solve this later.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+
+
+/// Stops the application.
+typedef Future<int> MultiDeviceAppsStopper(
```
This name seems odd. Why is this called `MultiDeviceAppsStopper`, when it actually stops the app on a single device? A better name would be `AppStopper` / `appStopper` / `stopApp()`.

kaiyuanw replied:
Renamed.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+}
+
+/// Runs driver tests.
+typedef Future<int> MultiDeviceTestRunner(List<String> testArgs);
```
This name should probably be kept as `TestRunner`.

kaiyuanw replied:
Good point.  Done.

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,375 @@
+  testRunner = runMultiDeviceTests;
+}
+
+Future<int> runMultiDeviceTests(List<String> testArgs) async {
```
Same here. just `runTests` would be better, because we're actually running a single test script. Consider directly calling the `runTests` defined in `driver.dart`, if possible.

kaiyuanw replied:
Agree.  I renamed it to `runTests`.

kaiyuanw committed:
Address issues for YoungSeok's comments, need to be reviewed again

kaiyuanw committed:
Remove FLUTTER_DEV in bin/flutter

yyoon commented on packages/flutter_tools/lib/src/commands/multi_drive.dart:
```diff
@@ -0,0 +1,366 @@
+      });
+      return newSpecs;
+    } on io.FileSystemException catch(e) {
+      print(e);
```
Use `printError` instead of `print`, and at least show some meaningful error messages (e.g. `"An error occurred while reading the test specs: $specsPath"`). You should avoid just printing out the exception as is.

kaiyuanw replied:
Fixed.

kaiyuanw committed:
Untrack bin/flutter and fix YoungSeok's comments

kaiyuanw committed:
Revert commit to keep bin/flutter in the repo

kaiyuanw committed:
Added kDefaultDebugPortBase and kDefaultPortRange

kaiyuanw committed:
Allow users to specify devices with device id or model name,
device id is optional so a multi-device test script can be
run on different places.  Automatically compute a one to one
mapping between specs and device so that each device specified
in specs file maps to a real device.  Automatically detect
available observatory port and assign it to a device for
monitoring.  Users are only required to specify a device
nickname and an application path to make flutter multi-drive
command works.  The rest attributes are optional, including
device id and model name.

kaiyuanw committed:
Support driver connection by device nickname.  pubspecs.yaml is modified to import kMultiDriveSpecsName from common.dart under flutter_tools package

kaiyuanw committed:
Add more documentations and allow users to specify debug-ports.  User specified debug-ports are considered first before the system tries to automatically assign debug ports to devices
